### ISSUE TITLE
cmake is unable to find PYTHON_LIBRARIES on macOS

### ISSUE THREAD

**Issue by Mrwerdo** [SFTtech/openage#994]
The problem: `cmake` is unable to find `PYTHON_LIBRARIES`
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
PYTHON_LIBRARIES
    linked by target "cmTC_c3988" in directory /Users/mrwerdo/Developer/openage/.bin/clang++-debug-Oauto-sanitize-none/CMakeFiles/CMakeTmp

CMake Error at buildsystem/modules/FindPython.cmake:181 (try_compile):
  Failed to configure test project build system.
Call Stack (most recent call first):
  buildsystem/HandlePythonOptions.cmake:6 (find_package)
  CMakeLists.txt:141 (include)
```
Python 3.6.4 is installed using homebrew.

On lines 166-169 of buildsystem/modules/FindPython.cmake, the variables considered are:
```
PYTHON_INCLUDE_DIRS="/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/include/python3.6m"
PYTHON_LIBRARY_DIR="/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/lib"
PYTHON_LIBPL="/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/lib/python3.6/config-3.6m-darwin"
PYTHON_LIBRARY_NAME="Python.framework/Versions/3.6/Python"
```

Replacing `PYTHON_LIBRARY_NAME` with `Python` results in `find_library` finding a system package (`/System/Library/Frameworks/Python.framework`, which has out of date versions of python: `2.3`, `2.5`, `2.6` and `2.7`). This appears to work, but later on it fails linking - missing various python functions.

Finally, **running `./configure -- "-DPYTHON_LIBRARIES=$(python3-config --prefix)/lib/libpython3.6.dylib"` actually works**. `python3-config --prefix` expands to `/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6`.

Perhaps `find_library` is failing to set `PYTHON_LIBRARIES` because `PYTHON_LIBRARY_DIR` has `lib` attached to the end of it?

**Comment by Mrwerdo** [comment]
It turns out after a lot of digging around, that:

1. `find_library(... NAMES "Python" ...)` finds the system's frameworks. This is kinda useful. If we set `CMAKE_FRAMEWORK_PATH=/usr/local/Frameworks` then this finds the Python 3 framework correctly. The problem with this is that later on the build fails during linking because a few`Py_*` functions are missing. cmake behaves differently when a framework is discovered [(see the paragraph about frameworks)](https://cmake.org/cmake/help/v3.8/command/find_library.html?highlight=i), I suspect this causes the linking problem.
2. `find_library(... NAMES $PYTHON_LIBRARY_NAME "libpython3.6.dylib" ...)` configures and builds correctly. The only issue with this is figuring out the version number.

Pursuing 1, we see that `CMAKE_FIND_LIBRARY_SUFFIXES=.tbd;.dylib;.so;.a`. If we add an empty suffix to the array, and force cmake to look for normal libraries first, cmake finds `lusr/local/Frameworks/Python.framework/Versions/3.6/Python`. The build works, which is a plus as well.

**Comment by TheJJ** [comment]
Okay. But what is the problem with our current approach that we ask a Python interpreter where its library is located and then use this path? This also guarantees to use the correct library version in case of multiple installed Pythons.
Is it just a search path problem or does `find_library` not discover the `libpythonx.x.dylib` even though it got the correct path?

**Comment by Mrwerdo** [comment]
Yeah, it's just `find_library` doesn't find `libpythonx.x.dylib`. Do you know of a clean way of discovering the version numbers?

**Comment by TheJJ** [comment]
On my Linux the library is called `/usr/lib/libpython3.6m.so`, which is exactly the same name scheme, just with the Linux-`.so` extension instead of `.dylib`. Can any other macOS user confirm this?

https://cmake.org/cmake/help/latest/command/find_library.html

Maybe the extension is not correctly configured to be `.dylib` for some reason? I can hardly immagine this being a problem to find a library at a known path with a standard name.

**Comment by Mrwerdo** [comment]
The problem is that `PYTHON_LIBRARY_DIR` + `PYTHON_LIBRARY_NAME` doesn’t exist. Specifically what’s messing up the discovery is that `PYTHON_LIBRARY_NAME` points to the framework’s product (relative to the framework), instead of the dylib. When you concatenation the two the paths, it doesn’t exist. If Python was telling us that the library name was `libpython3.6.dylib`, then everything would have worked as expected.

**Comment by TheJJ** [comment]
Hm. Currently we use `LDLIBRARY` from `distutils.sysconfig` to get the library name. Maybe that's just wrong.

Can you try see if the correct name is somewhere in another config variable?

```python
import distutils.sysconfig
import pprint

pprint.pprint({k: v for k, v in distutils.sysconfig.get_config_vars().items() if (isinstance(v, str) and 'python' in v.lower())})
```

You should see various config variables including our currently-used `LDLIBRARY`. The `"Python.framework/Versions/3.6/Python"` really shouldn't be this value. Maybe it's a faulty configuration for the macOS-build that homebrew is deploying?

**Comment by TheJJ** [comment]
How does the stock-`FindPython.cmake` find python in this case? Maybe we should try to borrow how they do it.

However, your last commit is the best workaround for now, if it works :smile:

If the symlink is actually to the file referenced by `LDLIBRARY`, then maybe cmake just has to be convinced that this is a shared library without the `.dylib` extension?

**Comment by Mrwerdo** [comment]
I’m not 100% sure on how the stock `FindPython.cmake` works, I’ll compare it to our one.

If you set `PYTHON_LIBRARY_NAME` to `Python` then `find_library` picks up various frameworks. This is because there is a special case to search for `NAME.framework` when on Darwin platforms. When I tried doing this, the project failed to link. In this case `PYTHON_LIBRARIES` is set to something like `/usr/local/opt/python/Frameworks/Python.framework`.

I managed to convince cmake that it’s a shared library. 😁 `find_library` checks the extensions of each file to make sure that they contain on of `CMAKE_FIND_LIBRARY_SUFFIXES` before determining it’s a library. It’s the only reason why it doesn’t find the shared library inside the framework. If you append an empty string to the variable, then `find_library` will find the shared library. This is more intrusive in practice, so I didn’t put it in the commit.

**Comment by TheJJ** [comment]
Okay. If you want, we can merge the workaround that way, but I'd be more happy if we found a better solution, maybe by inspiration from the original `FindPython`.

**Comment by Mrwerdo** [comment]
I had a quick look at the current implementation of [Find][1][Python][2] - it's pretty demotivating.

I think we should merge the workaround and leave this issue open. My reasoning is because it's currently mixed in with some other fixes, and that leaving the issue open means that other people are aware of the problem in the future.

[1]: https://github.com/Kitware/CMake/blob/master/Modules/FindPython.cmake
[2]: https://github.com/Kitware/CMake/blob/master/Modules/FindPython/Support.cmake

**Comment by TheJJ** [comment]
Looking at those files, we should likely improve our version and enhance it with new features such as `Component`s. Then again, it works pretty well right now, except that it doesn't prefer newer versions i think.
But yes, the workaround would be a first solution.

**Comment by TheJJ** [comment]
Fixed in #995.
